install-validator/internal/kubernetes/client.go

```diff
@@ -0,0 +1,113 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	ext "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/yaml"
+)
+
+type ClientInt interface {
+	GetCrd(name string) *v1.CustomResourceDefinition
+	ApplyCrd(crd v1.CustomResourceDefinition) error
+	ListCrds() error
+	ListResources(crd v1.CustomResourceDefinition) ([]interface{}, error)
+}
+
+type Client struct {
+	clientset *ext.Clientset
+	config    *rest.Config
+	crds      []v1.CustomResourceDefinition
+}
+
+func NewClient() (Client, error) {
+	c := Client{}
+	err := c.setConfig()
+	if err != nil {
+		return c, err
+	}
+	err = c.setClientSet()
+	if err != nil {
+		return c, err
+	}
+	return c, nil
+}
+
+func (c *Client) setClientSet() error {
+	clientset, err := ext.NewForConfig(c.config)
+
+	if err != nil {
+		return err
+	}
+	c.clientset = clientset
+
+	return nil
+}
+
+func (c *Client) setConfig() error {
+	config, err := rest.InClusterConfig()
+	if err == nil {
+		c.config = config
+		return nil
+	}
+
+	home, exists := os.LookupEnv("HOME")
+	if !exists {
+		home = "/root"
+	}
+	configPath := filepath.Join(home, ".kube", "config")
+	config, err = clientcmd.BuildConfigFromFlags("", configPath)
+	if err != nil {
+		return err
+	}
+	c.config = config
+	return nil
+}
+
+func (c *Client) GetCrd(name string) *v1.CustomResourceDefinition {
+	for _, s := range c.crds {
+		if s.Name == name {
+			return &s
+		}
+	}
+	return nil
+}
+
+func (c *Client) ApplyCrd(crd v1.CustomResourceDefinition) error {
+	oldCrd := c.GetCrd(crd.Name)
+	var err error
+	if oldCrd == nil {
```

Comment:
err should explicity check for error type to be "not exist"

Reply:
The app fetch all the crds to memory in initialization (because it needs them in few places) and the GetCrd functions just looks for the name in the map and return nil pointer if it cand find it. Therefore there is no "not exist" error - error is just initialized before switch, because both api calls are inside of the if statement.